src/alert_manager.cpp

```diff
@@ -152,54 +175,147 @@ namespace libtorrent
+	{
+		const int new_limit = m_queue_limit_requested;
+
+		// we can only resize if there is no other thread posting alerts
+		// and if no thread posted an alert since we popped them
+		if (new_limit < 0 || m_buffer_refs > 0 || m_queue_size > 0)
+			return;
+
+		// resize the queue
+		m_alerts = std::vector<boost::atomic<alert*>>(new_limit * 2);
+		for (int i = 0; i < new_limit * 2; i++)
+			m_alerts[i].store(NULL, boost::memory_order_relaxed);
+
+		m_queue_size_limit = new_limit;
+		m_queue_write_slot = -1;
+		m_queue_read_slot = 0;
+		m_queue_limit_requested = -1;
+	}
+
+	void alert_manager::get_all(std::vector<alert*>& alerts)
 	{
 		mutex::scoped_lock lock(m_mutex);
-		TORRENT_ASSERT(m_num_queued_resume <= m_alerts[m_generation].size());
 
+		if (m_queue_limit_requested != -1)
+		{
+			// if a buffer resize has been requested wait until all
+			// buffer references have been released to ensure that
+			// maybe_resize_buffer() succeeds
+			while (m_buffer_refs > 0) this_thread::yield();
+		}
+		else
+		{
+			// yield the cpu at least once before to ensure that
+			// this function cannot be called twice in the timespan
+			// between the alert contruction and it being placed in the
+			// queue (and available to us).
+			this_thread::yield();
+		}
+
+		//const int size_limit = m_queue_size_limit;
+
+		// release alerts in the pending deletes list
+		for (unsigned int i = 0; i < m_alerts_pending_delete.size(); i++)
+			delete m_alerts_pending_delete[i];
+		m_alerts_pending_delete.clear();
 		alerts.clear();
-		if (m_alerts[m_generation].empty()) return;
 
-		m_alerts[m_generation].get_pointers(alerts);
-		num_resume = m_num_queued_resume;
-		m_num_queued_resume = 0;
+		// after this libtorrent threads can begin work on posting
+		// new alerts if the queue was full.
+		const int n_alerts = m_queue_size.exchange(0);
 
-		// swap buffers
-		m_generation = (m_generation + 1) & 1;
-		// clear the one we will start writing to now
-		m_alerts[m_generation].clear();
-		m_allocations[m_generation].reset();
-	}
+		if (n_alerts == 0) return;
 
-	bool alert_manager::pending() const
-	{
-		mutex::scoped_lock lock(m_mutex);
-		return !m_alerts[m_generation].empty();
+		// pop n_alerts
+		for (int i = 0; i < n_alerts; i++)
+		{
+			// remove the alert and mark the ring buffer slot as free
+			alert * const alert = pop_alert();
+			TORRENT_ASSERT(alert != NULL);
+
+			// copy the alerts to the user's vector and to the
+			// pending delete list
+			m_alerts_pending_delete.push_back(alert);
+			alerts.push_back(alert);
+		}
+
+		// if a buffer resize is pending do it now
+		maybe_resize_buffer();
+
+		// free the storage for each thread
+		for (unsigned int i = 0; i < m_threads.size(); i++)
+		{
+			thread_state& ts = m_threads[i];
+
+			int gen = ts.youngest_generation
+				->load(boost::memory_order_relaxed);
+
+			// swap buffers
+			gen = (gen == TORRENT_ALERT_MANAGER_N_GENERATIONS - 1) ? 0 : (gen + 1);
+			ts.youngest_generation->store(gen);
```

Comment:
No, this is only updated by the user threads and it's done under a mutex. Every thread has 3 allocators so as long as it doesn't change twice between alert construction and it being placed on the queue there is no race. And we have a check in place to make sure it doesn't happen (`do_enqueue_alert()' will return false and set abort to true if it happens). I shouldn't happen under normal conditions but it can under stress testing.

Reply:
After looking at the code again, yes there was a race. The atomic store need to happen after we reset the oldest allocator, since it becomes the current allocator after the store.